docs/design/values.md

```diff
@@ -0,0 +1,1014 @@
+and `impl` search occur for the same type regardless of the expression category.
+But once a particular method or function is selected, an implicit conversion can
+occur from the original type to the representation type as part of the parameter
+or receiver type. In fact, this conversion is the _only_ operation that can
+occur for a value whose type has a customized value representation.
+
+The example above also demonstrates the fundamental tradeoff made by customizing
+the value representation of a type in this way. While it provides a great deal
+of control, it may result in some surprising limitations. Above, a method that
+is classically available on a C++ `const std::string&` like querying the
+capacity cannot be implemented with the customized value representation because
+it loses access to this additional state. Carbon allows type authors to make an
+explicit choice about whether they want to work with a restricted API and
+leverage a custom value representation or not.
+
+**Open question:** Beyond the specific syntax used where we currently have a
+placeholder `value_rep = T;`, we need to explore exactly what the best
+relationship is with the customization point. For example, should this syntax
+immediately forward declare `impl as ReferenceImplicitAs where .T = T`, thereby
+allowing an out-of-line definition of the `Convert` method and `... where _` to
+pick up the associated type from the syntax. Alternatively, the syntactic marker
+might be integrated into the `impl` declaration for `ReferenceImplicitAs`
+itself.
+
+### Polymorphic types
+
+Value expressions and value bindings can be used with
+[polymorphic types](/docs/design/classes.md#inheritance), for example:
+
+```carbon
+base class MyBase { ... }
+
+fn UseBase(b: MyBase) { ... }
+
+class Derived {
+  extend base: MyBase;
+  ...
+}
+
+fn PassDerived() {
+  var d: Derived = ...;
+  // Allowed to pass `d` here:
+  UseBase(d);
+}
+```
+
+This is still allowed to create a copy or to move, but it must not _slice_. Even
+if a copy is created, it must be a `Derived` object, even though this may limit
+the available implementation strategies.
+
+### Interop with C++ `const &` and `const` methods
+
+While value expressions cannot have their address taken in Carbon, they should
+be interoperable with C++ `const &`s and C++ `const`-qualified methods. This
+will in-effect "pin" some object (potentially a copy or temporary) into memory
+and allow C++ to take its address. Without supporting this, values would likely
+create an untenable interop ergonomic barrier. However, this does create some
+additional constraints on value expressions and a way that their addresses can
+escape unexpectedly.
+
+Despite interop requiring an address to implement, the address isn't guaranteed
+to be stable or useful or point back to some original object necessarily. The
+ability of the implementation to introduce copies or a temporary specifically
+for the purpose of the interop remains.
+
```

Comment:
Seems that this is missing the bit about why we aren't worried about C++ code casting away `const`. This is the text you had in `p2006.md`:

> It may seem risky to expose Carbon's value expressions to `const &` parameters
in this way, as C++ allows casting away `const`. However, in the absence of
`mutable` members, casting away `const` does not make it safe to _mutate_
through a `const &` parameter (or a `const`-qualified method). C++ objects can
be _declared_ `const`, and those objects are allowed to be implemented in
readonly memory and assumed to not mutate even if escaped. And `const &`
parameters can bind to these `const`-declared objects. As a consequence,
actually mutating through a `const &` parameter (perhaps after casting away
const) is a fundamentally unsafe operation in C++, and one that
[breaks in real implementations](https://cpp.compiler-explorer.com/z/KMhTondaK).
As a consequence, there is strong pressure that C++ code operate with `const &`
parameters in a way that will be fully compatible with Carbon's value
expressions.

(Wouldn't be bad to trim this down a bit. Need to go to bed now, though.)

Reply:
I tried to shorten it, but not sure it's actually shorter. Maybe it's a bit more accessible? 

Anyways, added this content, good call-out to make sure it is here.